File: rust/BubbleSort.rs
enum Sort {
    Ascending,
    Descending
}

fn bubble_sort<T: PartialOrd>(arr: &mut [T], sort: Sort) {
    let length = arr.len();

    match sort {
        Sort::Ascending => {
            for i in 0..length-1 {
                for j in 0..length-i-1 {
                    if arr[j] > arr[j+1] {
                        arr.swap(j, j+1);
                    }
                }
            }
        },
        Sort::Descending => {
            for i in 0..length-1 {
                for j in 0..length-i-1 {
                    if arr[j] < arr[j+1] {
                        arr.swap(j, j+1);
                    }
                }
            }
        }
    }
    
}

fn main() {
    let mut vec = vec![1,4,3,6,1,3,5,2];
    println!("Before sorting: {:?}", vec);
    bubble_sort(&mut vec, Sort::Ascending);
    println!("Sorted ascending: {:?}", vec);
    bubble_sort(&mut vec, Sort::Descending);
    println!("Sorted descending: {:?}", vec);
}

File: java/BubbleSort.java
import java.util.Arrays;

class BubbleSort {
  void bubbleSort(int array[]) {
    int size = array.length;
    
    // run loops two times: one for walking throught the array
    // and the other for comparison
    for (int i = 0; i < size - 1; i++)
      for (int j = 0; j < size - i - 1; j++)

        // To sort in descending order, change > to < in this line.
        if (array[j] > array[j + 1]) {

          // swap if greater is at the rear position
          int temp = array[j];
          array[j] = array[j + 1];
          array[j + 1] = temp;
        }
  }

  // usage
  public static void main(String args[]) {
    int[] data = { -2, 45, 0, 11, -9 };
    BubbleSort bs = new BubbleSort();
    bs.bubbleSort(data);
    System.out.println("Sorted Array:");
    System.out.println(Arrays.toString(data));
  }
}
